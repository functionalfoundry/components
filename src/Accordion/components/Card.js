/** @flow */
import React from 'react'
import Theme from 'js-theme'
import { Colors, Fonts, Spacing } from '@workflo/styles'

import Icon from '../../Icon'

type ContainerPropsT = {
  children: any,
  onCollapse: Function,
  onExpand: Function,
  /** When in uncontrolled mode, indicates the initial expanded state */
  isInitiallyExpanded: boolean,
  /**
   * Setting this prop puts the Card into "controlled" mode and also controls
   * the expanded state in this mode.
   */
  isExpanded: boolean,
  title: string,
  theme: Object,
}

type ContainerStateT = {
  /** Indicates whether the Card is in controlled or uncontrolled mode */
  isControlled: boolean,
  isExpanded: boolean,
}

/**
 * Card component for rendering an expandable/collapsable portion of an Accordion.
 * Can be used in both controlled and uncontrolled modes, determined by whether the
 * isExpanded prop is specified. Can be used standalone or with an Accordion parent
 * in which case the Accordion parent will take ownership of controlling and managing
 * it's Card children.
 */
class CardContainer extends React.Component {
  props: ContainerPropsT
  state: ContainerStateT

  constructor(props: ContainerPropsT) {
    const { isExpanded, isInitiallyExpanded } = props
    super(props)
    this.state = {
      isControlled: this.getIsControlled(props),
      isExpanded: this.getIsControlled(props) ? isExpanded : isInitiallyExpanded,
    }
  }

  getIsControlled = ({ isExpanded }: any) =>
    isExpanded !== null && typeof isExpanded !== 'undefined'

  handleToggle = () => {
    if (!this.state.isControlled) {
      this.setState(({ isExpanded }) => ({ isExpanded: !isExpanded }))
    }
  }

  render() {
    return (
      <ThemedCard
        {...this.props}
        isExpanded={this.state.isExpanded}
        onToggle={this.handleToggle}
      />
    )
  }
}

type PropsT = {
  children: any,
  isExpanded: boolean,
  onToggle: Function,
  title: string,
  theme: Object,
}

const Card = ({ children, isExpanded = true, onToggle, title, theme }: PropsT) => (
  <div {...theme.cardContainer}>
    <div {...theme.cardTitleContainer}>
      <div {...theme.cardExpandIcon} onClick={onToggle}>
        {isExpanded
          ? <Icon name="minus" stroke={Colors.grey300} />
          : <Icon name="add" stroke={Colors.grey300} />}
      </div>
      <div {...theme.cardTitle}>{title}</div>
    </div>
    <div {...theme.cardContent}>
      {children}
    </div>
  </div>
)

const defaultTheme = ({ isExpanded }) => ({
  cardContainer: {
    ...Fonts.small,
    borderColor: Colors.grey900,
    borderStyle: 'solid',
    borderWidth: 1,
    flexGrow: 1,
  },
  cardContent: isExpanded
    ? {
        overflow: 'scroll',
      }
    : {
        overflow: 'hidden',
        height: 0,
      },
  cardTitle: {
    paddingLeft: Spacing.tiny,
  },
  cardTitleContainer: {
    ...Fonts.small,
    display: 'flex',
    flexDirection: 'row',
  },
})

const ThemedCard = Theme('Card', defaultTheme)(Card)

export default CardContainer
