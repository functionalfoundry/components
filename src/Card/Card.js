/* @flow */
import React from 'react'
import Theme from 'js-theme'
import { Corners, Spacing, Shadows } from '@workflo/styles'
import View from '../View'

type PropsT = {
  children?: React.Children,
  size: string,
  testId?: string,
  theme: Object,
  flush: boolean,
  floating: boolean,
}

const defaultProps = {
  size: 'base',
  theme: {},
  flush: false,
  floating: false,
}

const Card = ({
  children,
  theme,
  flush, // eslint-disable-line no-unused-vars
  floating, // eslint-disable-line no-unused-vars
  testId, // eslint-disable-line no-unused-vars
  ...props
}: PropsT) => (
  <View {...props} {...theme.card}>
    {children}
  </View>
)

Card.defaultProps = defaultProps

const getFloating = ({ floating }: PropsT) => {
  if (floating) {
    return Shadows.base
  }
  return {}
}

const getSize = ({ size }: PropsT) => {
  switch (size) {
    case 'tiny':
      return {
        width: 100,
        height: 60,
      }
    case 'small':
      return {
        width: 200,
        height: 180,
      }
    case 'base':
      return {
        width: 280,
        height: 240,
      }
    case 'large':
      return {
        width: 320,
        height: 260,
      }
    case 'huge':
      return {
        width: 400,
        height: 400,
      }
    default:
      return {}
  }
}

const defaultTheme = (props: PropsT) => ({
  card: {
    ...Corners.rounded,
    ...getSize(props),
    ...getFloating(props),
    ...getPadding(props.flush),
    backgroundColor: 'white',
  },
})

const getPadding = (flush: boolean) => {
  if (flush) {
    return {
      padding: 0,
    }
  }
  return {
    padding: Spacing.small,
  }
}

const ThemedCard = Theme('Card', defaultTheme)(Card)
export default ThemedCard
