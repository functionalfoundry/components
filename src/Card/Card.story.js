import React from 'react'
import { storiesOf } from '@kadira/storybook'
import {
  Colors,
  Fonts,
  Spacing,
} from '@workflo/styles'
import Card from '.'
import View from '../View'

type CardViewPropsT = {
  fontStyle: Object,
  cardProps: Object,
  label: string,
}

const CardView = ({
  fontStyle,
  cardProps,
  label,
}: CardViewPropsT) => (
  <View>
    <View
      style={{
        ...fontStyle,
      }}
    >
      {label}
    </View>
    <Card
      {...cardProps}
      style={{
        backgroundColor: 'white',
        margin: 20,
      }}
    />
  </View>
)

storiesOf('Cards', module)
  .add('Sizes', () => (
    <div
      style={{
        backgroundColor: Colors.grey1200,
        padding: Spacing.base,
        display: 'flex',
        flex: 1,
        flexDirection: 'column',
      }}
    >
      <CardView
        key={'tiny'}
        fontStyle={Fonts.base}
        cardProps={{
          size: 'tiny',
        }}
        label={'tiny'}
      />
      <CardView
        key={'small'}
        fontStyle={Fonts.base}
        cardProps={{
          size: 'small',
        }}
        label={'small'}
      />
      <CardView
        key={'base'}
        fontStyle={Fonts.base}
        cardProps={{
          size: 'base',
        }}
        label={'base'}
      />
      <CardView
        key={'large'}
        fontStyle={Fonts.base}
        cardProps={{
          size: 'large',
        }}
        label={'large'}
      />
      <CardView
        key={'huge'}
        fontStyle={Fonts.base}
        cardProps={{
          size: 'huge',
        }}
        label={'huge'}
      />
    </div>
  ))
  .add('Floating', () => (
    <div
      style={{
        backgroundColor: Colors.grey1200,
        padding: Spacing.base,
        display: 'flex',
        flex: 1,
      }}
    >
      <CardView
        key={'floating'}
        fontStyle={Fonts.base}
        cardProps={{
          floating: true,
        }}
        label={'Floating'}
      />
    </div>
  ))
