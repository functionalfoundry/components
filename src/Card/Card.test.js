import React from 'react'
import renderer from 'react-test-renderer'
import Card from '.'

test('Renders without props', () => {
  const card = renderer.create(
    <Card />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders children', () => {
  const card = renderer.create(
    <Card>
      <div>{'Child'}</div>
    </Card>
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders tiny size', () => {
  const card = renderer.create(
    <Card
      size='tiny'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders small size', () => {
  const card = renderer.create(
    <Card
      size='small'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders base size', () => {
  const card = renderer.create(
    <Card
      size='base'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders large size', () => {
  const card = renderer.create(
    <Card
      size='large'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders huge size', () => {
  const card = renderer.create(
    <Card
      size='huge'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders unknown size', () => {
  const card = renderer.create(
    <Card
      size='unknown'
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})

test('Renders floating', () => {
  const card = renderer.create(
    <Card
      floating
    />
  ).toJSON()
  expect(card).toMatchSnapshot()
})
