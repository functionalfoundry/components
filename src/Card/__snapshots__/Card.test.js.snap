exports[`test Renders base size 1`] = `
<div
  className="view_1ps7m7i card_fwvbvf"
  data-test-id={undefined}
  size="base" />
`;

exports[`test Renders children 1`] = `
<div
  className="view_1ps7m7i card_fwvbvf"
  data-test-id={undefined}
  size="base">
  <div>
    Child
  </div>
</div>
`;

exports[`test Renders floating 1`] = `
<div
  className="view_1ps7m7i card_e3ecr9"
  data-test-id={undefined}
  size="base" />
`;

exports[`test Renders huge size 1`] = `
<div
  className="view_1ps7m7i card_12yu5ad"
  data-test-id={undefined}
  size="huge" />
`;

exports[`test Renders large size 1`] = `
<div
  className="view_1ps7m7i card_10ez84b"
  data-test-id={undefined}
  size="large" />
`;

exports[`test Renders small size 1`] = `
<div
  className="view_1ps7m7i card_m0hiek"
  data-test-id={undefined}
  size="small" />
`;

exports[`test Renders tiny size 1`] = `
<div
  className="view_1ps7m7i card_u5y1ol"
  data-test-id={undefined}
  size="tiny" />
`;

exports[`test Renders unknown size 1`] = `
<div
  className="view_1ps7m7i card_16xse8r"
  data-test-id={undefined}
  size="unknown" />
`;

exports[`test Renders without props 1`] = `
<div
  className="view_1ps7m7i card_fwvbvf"
  data-test-id={undefined}
  size="base" />
`;
